File: frontend/types/analytics.ts
// Mirrors backend/app/schemas/analytics.py (IndustryAnalyticsResponse).
// Every metric is a live aggregate of the authenticated Industry
// account's OWN records — there is no analytics table.
//
// Historical scope: the database records only each row's creation date
// (and applications' apply date). It does NOT record when a status
// changed, so `timeline` is strictly "records created per month" and
// there is no stage-conversion-over-time metric. `historical_note`
// carries this caveat in the payload itself.

import type { ApplicationStatus } from "@/types/application";

export interface AnalyticsKpis {
  opportunities_total: number;
  opportunities_published: number;
  applications_total: number;
  shortlisted: number;
  interviews_total: number;
  interviews_upcoming: number;
  selected: number;
  collaborations_total: number;
  collaborations_active: number;
}

export interface OpportunityTypeBreakdown {
  opportunity_type: string;
  total: number;
  published: number;
}

export interface StatusCount {
  status: string;
  count: number;
}

export interface InterviewMetrics {
  total: number;
  scheduled: number;
  completed: number;
  cancelled: number;
  upcoming: number;
}

export interface TopOpportunity {
  id: string;
  title: string;
  opportunity_type: string;
  application_count: number;
}

export interface TimePoint {
  /** "YYYY-MM" */
  period: string;
  opportunities_created: number;
  applications_received: number;
}

export interface IndustryAnalytics {
  generated_at: string;
  kpis: AnalyticsKpis;
  funnel_counts: Record<ApplicationStatus, number>;
  funnel_total: number;
  application_status_distribution: StatusCount[];
  opportunity_breakdown: OpportunityTypeBreakdown[];
  interview_metrics: InterviewMetrics;
  top_opportunities: TopOpportunity[];
  timeline: TimePoint[];
  historical_note: string;
  interviews_available: boolean;
}

export const OPPORTUNITY_TYPE_LABELS: Record<string, string> = {
  INTERNSHIP: "Internships",
  JOB: "Jobs",
  PROJECT: "Projects",
  TRAINING: "Training",
  WORKSHOP: "Workshops",
  MENTORSHIP: "Mentorship",
  UNKNOWN: "Other",
};

/** One-line definition for every metric shown on the dashboard, so each
 * number is self-explanatory (spec: "Every displayed metric must have a
 * clear definition"). */
export const METRIC_DEFINITIONS = {
  opportunities:
    "Every internship, job, project, training, workshop and mentorship posting on your account, in any status.",
  published: "Postings currently PUBLISHED and visible to students.",
  applications: "Every application ever submitted to one of your internship or job postings.",
  shortlisted: "Applications currently at the Shortlisted stage.",
  interviews: "Interview records you have created (scheduled, completed or cancelled).",
  upcomingInterviews: "Scheduled interviews whose date/time has not yet passed.",
  selected: "Applications you have marked Selected.",
  collaborations: "Collaboration proposals you have created, in any lifecycle state.",
  activeCollaborations: "Collaborations currently ACTIVE.",
  funnel:
    "Where your applications sit right now across the recruitment pipeline. This is a current snapshot, not movement over time.",
  statusDistribution: "Count of your applications in each status, including exits (rejected/withdrawn).",
  breakdown: "How many postings you have per module, and how many of each are published.",
  interviewMetrics: "Your interview records grouped by status, plus how many are still upcoming.",
  topOpportunities: "Your postings ranked by how many applications they have received.",
  timeline:
    "Postings created and applications received per month, over the last 6 months. Based on creation dates only.",
} as const;

File: backend/app/schemas/analytics.py
"""Pydantic schemas for Industry Analytics.

Every metric here is computed live from the caller's OWN records
(applications, interviews, the six opportunity modules, collaborations) --
scoped by industry_id, through a user-scoped RLS client. There is no
analytics table and no stored aggregate.

Historical scope is deliberately limited to what the data can honestly
support: the schema records each row's `created_at` (and applications'
`applied_at`), so "records created over time" is real. It does NOT record
WHEN a status changed -- only the current status plus a single
`updated_at` -- so point-in-time funnel movement (when a candidate was
shortlisted, selected, ...) and any stage-conversion trend over time are
not available. `IndustryAnalyticsResponse.historical_note` states this in
the payload itself.
"""

from pydantic import BaseModel


class AnalyticsKpis(BaseModel):
    """Current-state headline counts. Every value is a count of the
    caller's own rows as they stand right now."""

    opportunities_total: int
    opportunities_published: int
    applications_total: int
    shortlisted: int
    interviews_total: int
    interviews_upcoming: int
    selected: int
    collaborations_total: int
    collaborations_active: int


class OpportunityTypeBreakdown(BaseModel):
    """One opportunity module: how many the caller has, and how many are
    live (PUBLISHED)."""

    opportunity_type: str
    total: int
    published: int


class StatusCount(BaseModel):
    status: str
    count: int


class InterviewMetrics(BaseModel):
    total: int
    scheduled: int
    completed: int
    cancelled: int
    upcoming: int


class TopOpportunity(BaseModel):
    """A posting ranked by how many applications it has received."""

    id: str
    title: str
    opportunity_type: str
    application_count: int


class TimePoint(BaseModel):
    """One calendar month. `opportunities_created` counts postings created
    that month across all six modules; `applications_received` counts
    applications whose `applied_at` falls in that month. Both are
    creation-date facts -- never inferred status history."""

    period: str  # "YYYY-MM"
    opportunities_created: int
    applications_received: int


class IndustryAnalyticsResponse(BaseModel):
    generated_at: str
    kpis: AnalyticsKpis
    # Per-status application counts (every status present, 0 when none) --
    # identical shape to GET /api/v1/applications/summary so the frontend
    # recruitment funnel renders it directly.
    funnel_counts: dict[str, int]
    funnel_total: int
    application_status_distribution: list[StatusCount]
    opportunity_breakdown: list[OpportunityTypeBreakdown]
    interview_metrics: InterviewMetrics
    top_opportunities: list[TopOpportunity]
    timeline: list[TimePoint]
    historical_note: str
    # True only when the interviews table (migration 030) is present.
    # False -> interview_metrics is all zeros and the note explains why.
    interviews_available: bool
